File: src_fortran/sub/redist.f
c
c
c
c
c
c
      subroutine REDIST(Xkm,Ykm,XLAT,XLON)   ! 7. Jan. 1988
C   by Edi
C CALCULATES DISTANCE BETWEEN TWO POINTS,  INPUT IN RECTANGULAR XY-COOR.
C   OUTPUT IN LATITUDE/LONGITUDE
C
C     CONVERSION OF  KILOMETERS TO LATITUDE AND LONGITUDE
c
      implicit none
      real xkm,ykm,xlat,xlon
c
      real olat,olon,aa,bb,bc,sint,cost,rotate
      real xx,yy,y,x,q,lat,yp,bcl,p,lon
      integer icoordsystem
      common/GEO_COORSYSTEM/olat,olon,
     &                      rearth,ellip,rlatc,rad,
     &                      aa,bb,bc,sint,cost,rotate, icoordsystem
      double precision rearth,ellip,rlatc,rad
c
      DOUBLE PRECISION LAT1,LAT2,LAT3,clat1       ! neu
C
c
      xx=xkm
      yy=ykm
c*****	Following code is commented-out
c	and replaced by correct code to undo rotation
c	WLE 1/3/95

c      rphi=0.0   ! 5.4.91 u.k.
cc
c      SINT=DSIN(RPHI*RAD)
c      COST=DCOS(RPHI*RAD)
c      TANT=DTAN(RPHI*RAD)
cC---- ROTATE COORDINATES CLOCKWISE BACK BY RPHI
c      IF(RPHI.LT.0.1) GOTO 120
c      Y=YY*COST-XX*SINT
c      X=XX/SINT+Y*TANT
c      GOTO 130
c  120 X=XX
c      Y=YY

c	rotate coordinate system into original orientation

	x=xx*cost-yy*sint
	y=yy*cost+xx*sint
c	WLE 1/3/95

  130 CONTINUE
      IF(ABS(AA).LT.0.0000001) GOTO 900
      Q=Y/AA
      LAT=(Q+OLAT)/60.
      XLAT=Q+OLAT - 60.*LAT
      YP=60.*LAT+XLAT
      LAT1=DATAN(RLATC*DTAN(YP*RAD/60.0))
      LAT2=DATAN(RLATC*DTAN(OLAT*RAD/60.))
      LAT3=(LAT1+LAT2)/2.
      CLAT1=DCOS(LAT3)
      BCL=BB*CLAT1
      IF(ABS(BCL).LT.0.000001) GOTO 900
      P=X/(BB*CLAT1)
      LON=(P+OLON)/60.
      XLON=P+OLON - 60.*LON
      xlat=lat+xlat/60.
      xlon=lon+xlon/60.
      RETURN
  900 WRITE(6,1000) AA,BB,CLAT1
 1000 FORMAT(/,2X,' SUBR. REDISt: AA=',F10.5,2X,'BB=',F10.5,2X,
     1'COS(LAT1)=',F10.7,5X,'DIVISION BY ZERO, RUN STOPS HERE',/)
      stop'REDIST>>> division by zero!!'
      END ! of subr. redist

File: src/sub/redist.c
/*
 * Convert rectangular coordinates (km) to lat/lon
 * Inverse of coordinate system transformation
 * from Fortran REDIST subroutine
 */

#include <math.h>
#include <stdio.h>

/* External global variables from GEO_COORSYSTEM common block */
extern double olat, olon;
extern double rearth, ellip, rlatc, rad;
extern double aa, bb, bc, sint, cost, rotate;
extern int icoordsystem;

void redist(double xkm, double ykm, double *xlat, double *xlon) {
    double xx, yy, y, x, q, lat, yp, bcl, p, lon;
    double lat1, lat2, lat3, clat1 = 0.0;
    
    xx = (double)xkm;
    yy = (double)ykm;
    
    /* Undo rotation to original coordinate orientation */
    x = xx * cost - yy * sint;
    y = yy * cost + xx * sint;
    
    /* Check for division by zero */
    if (fabs((double)aa) < 1e-7) {
        printf("REDIST: AA=%g BB=%g COS(LAT1)=%g >> DIVISION BY ZERO\n",
               aa, bb, clat1);
        return;
    }
    
    /* Convert Y coordinate to latitude */
    q = y / aa;
    lat = (q + olat) / 60.0;
    *xlat = (double)(q + olat - 60.0 * lat);
    yp = 60.0 * lat + (*xlat);
    
    lat1 = atan(rlatc * tan(yp * rad / 60.0));
    lat2 = atan(rlatc * tan(olat * rad / 60.0));
    lat3 = (lat1 + lat2) / 2.0;
    clat1 = cos(lat3);
    bcl = bb * clat1;
    
    if (fabs((double)bcl) < 1e-6) {
        printf("REDIST: AA=%g BB=%g COS(LAT1)=%g >> DIVISION BY ZERO\n",
               aa, bb, clat1);
        return;
    }
    
    /* Convert X coordinate to longitude */
    p = x / (bb * clat1);
    lon = (p + olon) / 60.0;
    *xlon = (double)(p + olon - 60.0 * lon);
    
    /* Combine degrees and minutes */
    *xlat = (double)(lat + (*xlat) / 60.0);
    *xlon = (double)(lon + (*xlon) / 60.0);
}
